src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -809,31 +807,55 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
+            int currentBestIndex;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (currentBestIndex = 0; currentBestIndex < candidates.Count; currentBestIndex++)
+            {
+                if (candidates[currentBestIndex].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            for (int index = currentBestIndex + 1; index < candidates.Count; index++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
                 {
                     continue;
                 }
 
-                // Is this applicable operator better than every other applicable method?
-                for (int j = 0; j < i; ++j)
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
                 {
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
-                    {
-                        continue;
-                    }
-                    var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
-                    {
-                        candidates[j] = candidates[j].Worse();
-                    }
-                    else if (better == BetterResult.Right)
-                    {
-                        candidates[i] = candidates[i].Worse();
-                    }
+                    candidates[index] = candidates[index].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                    currentBestIndex = index;
+                }
+            }
+
+            for (int index = 0; index < candidates.Count; index++)
+            {
+                if (index == currentBestIndex || candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    candidates[index] = candidates[index].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                    break;
+                }
+                else
+                {
+                    break;
```

Comment:
> I'm still wondering why all tests passed (including the one I provided)!

We cannot rely on existing tests to catch errors in new algorithm. Given old implementation, we didn't have to test scenarios that are interesting for the new algorithm.

> I'd be nice if you could suggest some test ideas that would fail with this code.

I am just reading the code and trying to poke holes in it by thinking about potential situations the code might have to deal with. It would be rather hard to come up with specific scenario that would give just the "right" relationship between candidates and their order in the set. Often, correctness of an algorithm can be proven by providing logical reasoning to support the claim.

I recommend starting with trying to answer the question that started this thread: "Is it possible that, if we continue, we could downgrade current best to Worse?" Provide a logical reasoning behind the answer.